3 internal PaymentRequest tests are crashing in debug

cdumez opened the pull request:
#### f69976c08fd705f03ff8d384f63161465f129de9
<pre>
3 internal PaymentRequest tests are crashing in debug
<a href="https://bugs.webkit.org/show_bug.cgi?id=258988">https://bugs.webkit.org/show_bug.cgi?id=258988</a>
rdar://105024815

Reviewed by Devin Rousso.

The tests in question call abort() on the PaymentRequest from unexpected places,
and our code was missing checks in several places to make sure there is still
an active payment session, causing crashes.

To address the issue, we now check if we still have an active payment session in
more places, in particular after calling functions that may run JavaScript, since
JS may abort the request at any point.

This change is covered by those 3 internal tests, which I&apos;ll unskip.

* Source/WebCore/Modules/applepay/paymentrequest/ApplePayPaymentHandler.cpp:
(WebCore::ApplePayPaymentHandler::merchantValidationCompleted):
(WebCore::ApplePayPaymentHandler::shippingAddressUpdated):
(WebCore::ApplePayPaymentHandler::retry):
* Source/WebCore/Modules/paymentrequest/PaymentRequest.cpp:
(WebCore::PaymentRequest::abortWithException):
* Source/WebCore/testing/MockPaymentCoordinator.cpp:
(WebCore::MockPaymentCoordinator::dispatchIfShowing):

Canonical link: <a href="https://commits.webkit.org/265867@main">https://commits.webkit.org/265867@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/04cac3cd23fe66e4062a4522c3a334831717f222

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/12085 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/12431 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/12733 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/13831 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/11664 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/12116 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/14847 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/12445 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/14357 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/12249 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/13062 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/10234 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/14246 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/10350 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/10972 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/18097 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/11430 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/11132 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/14308 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/11619 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/9575 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/10832 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/2966 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/15158 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/11470 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

aprotyas reviewed (approved):
This looks good to me!

Should we consider adding similar checks elsewhere where we run JS? Like in `ApplePaySession::completeMerchantValidation()`

```cpp
    // Session may not be active after this?
    auto merchantSession = PaymentMerchantSession::fromJS(state, asObject(merchantSessionValue), errorMessage);
    if (!merchantSession) {
        window.printErrorMessage(errorMessage);
        return Exception { InvalidAccessError };
    }
```

dcrousso commented on Source/WebCore/Modules/applepay/paymentrequest/ApplePayPaymentHandler.cpp:
```diff
@@ -663,6 +668,12 @@ ExceptionOr<void> ApplePayPaymentHandler::shippingAddressUpdated(Vector<RefPtr<A
+
+    // firstApplicableModifier() may run JS, which may abort the request so we need
+    // to make sure we still have an active session.
+    if (!paymentCoordinator().hasActiveSession())
```
is this one actually a problem?  IIRC we serialize the `modifiers` to a string when first ingesting theme, so all `firstApplicableModifier()` does is parse it back into a JSON object and then convert that to the relevant IDL/C++ dictionary (i.e. there shouldn't be any opportunity for user JS to run when doing this since it's JSON parsed from a string)

if it is an issue, should we be doing this check for all the other places that `firstApplicableModifier()` is used (e.g. `ApplePayPaymentHandler::shippingOptionUpdated`)?

cdumez replied:
Yes, this one definitely was needed to address the crashes. I believe it is this call inside `firstApplicableModifier()` that can run script:
```
auto applePayModifier = convertDictionary<ApplePayModifier>(lexicalGlobalObject, WTFMove(data));
```

I'll look at other call sites for firstApplicableModifier(). They likely need post validation too.

cdumez replied:
I added checks after the other `firstApplicableModifier()` call sites.

dcrousso replied:
right but the `data` is coming from a `JSONParse` from a string

```
data = JSONParse(&lexicalGlobalObject, serializedModifierData[i]);
```

so there shouldn't be any way for user JS to run from that.  the only issue i can see is if the given JSON doesnt actually match `ApplePayModifier`, but that should already throw an exception, so i dont see how we can successfully return from `ApplePayPaymentHandler::firstApplicableModifier` and yet somehow also abort.  can you elaborate on the scenario where this is crashing?

cdumez replied:
You're right, I'll double check and report back. Maybe I can move the hasActiveSession() check earlier.

cdumez replied:
It was computeErrors() at the call site which could execute JavaScript so I'm moving the check after computeErrors() calls instead.

dcrousso commented on Source/WebCore/testing/MockPaymentCoordinator.cpp:
```diff
@@ -101,7 +101,9 @@ MockPaymentCoordinator::~MockPaymentCoordinator()
 void MockPaymentCoordinator::dispatchIfShowing(Function<void()>&& function)
 {
-    ASSERT(showCount > hideCount);
+    if (showCount <= hideCount)
+        return;
```
i would think that trying to do something after it's been hidden would be a bug, no?

or is this now erroneous given the above changes?

cdumez replied:
But nothing prevents the tests from calling it after the session has been aborted, via `internals.mockPaymentCoordinator.cancelPayment();`

dcrousso commented on Source/WebCore/Modules/paymentrequest/PaymentRequest.cpp:
```diff
@@ -440,6 +440,10 @@ void PaymentRequest::show(Document& document, RefPtr<DOMPromise>&& detailsPromis
 
 void PaymentRequest::abortWithException(Exception&& exception)
 {
+    // If state is "closed", then the request has already been aborted.
+    if (m_state == State::Closed)
+        return;
```
(forgot to hit "Add review comment" 😅)

are we now allowing for an abort more than once?  personally id rather us find the places where we're trying to abort when already aborted and prevent those scenarios from happening, especially since `PaymentRequest::closeActivePaymentHandler` is pretty inconsequential (i.e. it just resets some member variables)

cdumez replied:
We have code that calls an internal function, which may return an exception, it is returns an exception, we call `abortWithException()` with that exception. The issue is that the internal function may return an exception because we've already aborted. In this case, we would end up calling `abortWithException()` on something that is already aborted.

At first, I had the `m_state != State::Closed` at the couple of call sites that call abortWithException() with a "forwarded" exception but then I decided to simplify the code and move the check inside abortWithException().

An example of such call site is:
```
        auto exception = activePaymentHandler->merchantValidationCompleted(m_merchantSessionPromise->result());
        if (exception.hasException()) {
            abortWithException(exception.releaseException());
            return;
        }
```

Which we would have updated to:
```
        auto exception = activePaymentHandler->merchantValidationCompleted(m_merchantSessionPromise->result());
        if (exception.hasException()) {
            if (m_state != State::Closed)
                abortWithException(exception.releaseException());
            return;
        }
```

Please let me know what you think now that I've clarified.

dcrousso replied:
i see.  i would almost prefer the latter so that we could then know exactly where this kind of double abort can happen and handle cases where it's unexpected, but i also see the benefit of having it be generalized inside `PaymentRequest::abortWithException`

cdumez committed:
3 internal PaymentRequest tests are crashing in debug
https://bugs.webkit.org/show_bug.cgi?id=258988
rdar://105024815

Reviewed by Devin Rousso.

The tests in question call abort() on the PaymentRequest from unexpected places,
and our code was missing checks in several places to make sure there is still
an active payment session, causing crashes.

To address the issue, we now check if we still have an active payment session in
more places, in particular after calling functions that may run JavaScript, since
JS may abort the request at any point.

This change is covered by those 3 internal tests, which I'll unskip.

* Source/WebCore/Modules/applepay/paymentrequest/ApplePayPaymentHandler.cpp:
(WebCore::ApplePayPaymentHandler::merchantValidationCompleted):
(WebCore::ApplePayPaymentHandler::shippingAddressUpdated):
(WebCore::ApplePayPaymentHandler::retry):
* Source/WebCore/Modules/paymentrequest/PaymentRequest.cpp:
(WebCore::PaymentRequest::abortWithException):
* Source/WebCore/testing/MockPaymentCoordinator.cpp:
(WebCore::MockPaymentCoordinator::dispatchIfShowing):

Canonical link: https://commits.webkit.org/265867@main
